clang/include/clang/Basic/Attr.td

```diff
@@ -1705,20 +1705,17 @@ def SYCLIntelMaxWorkGroupSize : InheritableAttr {
   let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
   let Subjects = SubjectList<[Function], ErrorDiag>;
   let AdditionalMembers = [{
-    std::optional<llvm::APSInt> getXDimVal() const {
-      if (const auto *CE = dyn_cast<ConstantExpr>(getXDim()))
-        return CE->getResultAsAPSInt();
-      return std::nullopt;
+    unsigned getXDimVal() const {
+      assert(isa<ConstantExpr>(getXDim()));
```

Comment:
NIT: I don't think you need to assert on `isa<T>` here, [`cast<T>` will do it](https://github.com/jchlanda/llvm/blob/sycl/llvm/include/llvm/Support/Casting.h#L577) for you (here and in `Y`, `Z`).

Reply:
Very true, in fact, `you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator.`[1] - though I see it used in (mostly older) LLVM code. If it gives a more meaningful assertion message to the user than the `cast` assertion I could maybe see this pattern's utility It's also not clear from the guidance whether this is refers to conditional cases, like `if (isa<X>(V)) { cast<X>(V); }` or whether it means ever, like ever ever.

As it stands I haven't provided a useful assertion message so you're right to call it out, thanks!

Looking around similar attribute-handling code in clang, I suspect we are doing something slightly off as it's only really SYCL attributes that have casts to "ConstantExpr". I wonder if this is due to us wanting to use attributes with templated values? I'd have to look into it deeper. As it stands, enough SYCL attributes just blindly `cast` so I'll do it here too, for consistency.

[1] https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates

Reply:
done!